river/tree/last_classifier.py

```diff
@@ -0,0 +1,404 @@
+            if self._leaf_prediction == self._MAJORITY_CLASS:
+                return LeafMajorityClassWithDetector(initial_stats, depth, self.splitter, self.change_detector.clone())
+            elif self._leaf_prediction == self._NAIVE_BAYES:
+                return LeafNaiveBayesWithDetector(initial_stats, depth, self.splitter, self.change_detector.clone())
+            else:  # Naives Bayes Adaptive (default)
+                return LeafNaiveBayesAdaptiveWithDetector(initial_stats, depth, self.splitter, self.change_detector.clone())
+        else:
+            split_criterion = self._new_split_criterion()
+            if self._leaf_prediction == self._MAJORITY_CLASS:
+                return LeafMajorityClassWithDetector(initial_stats, depth, self.splitter, self.change_detector.clone(), split_criterion)
+            elif self._leaf_prediction == self._NAIVE_BAYES:
+                return LeafNaiveBayesWithDetector(initial_stats, depth, self.splitter, self.change_detector.clone(), split_criterion)
+            else:  # Naives Bayes Adaptive (default)
+                return LeafNaiveBayesAdaptiveWithDetector(initial_stats, depth, self.splitter, self.change_detector.clone(), split_criterion)
+    
+
+    def _new_split_criterion(self):
+        if self._split_criterion == self._GINI_SPLIT:
+            split_criterion = GiniSplitCriterion(self.min_branch_fraction)
+        elif self._split_criterion == self._INFO_GAIN_SPLIT:
+            split_criterion = InfoGainSplitCriterion(self.min_branch_fraction)
+        elif self._split_criterion == self._HELLINGER_SPLIT:
+            if not self.track_error:
+                raise ValueError("The Heillinger distance cannot estimate the purity of a single distribution.\
+                                 Use another split criterion or set track_error to True")
+            split_criterion = HellingerDistanceCriterion(self.min_branch_fraction)
+        else:
+            split_criterion = InfoGainSplitCriterion(self.min_branch_fraction)
+
+        return split_criterion
+
+    def _attempt_to_split(self, leaf: HTLeaf, parent: DTBranch, parent_branch: int, **kwargs):
+        """Attempt to split a leaf.
+
+        If the samples seen so far are not from the same class then:
+
+        1. Find split candidates and select the top 1.
+        2. If the top1 is greater than zero:
+           3.1 Replace the leaf node by a split node (branch node).
+           3.2 Add a new leaf node on each branch of the new split node.
+           3.3 Update tree's metrics
+
+        Optional: Disable poor attributes. Depends on the tree's configuration.
+
+        Parameters
+        ----------
+        leaf
+            The leaf to evaluate.
+        parent
+            The leaf's parent.
+        parent_branch
+            Parent leaf's branch index.
+        kwargs
+            Other parameters passed to the new branch.
+        """
+        if not leaf.observed_class_distribution_is_pure():  # type: ignore
```

Comment:
I was wondering. Rather than using the "purity" method you introduced for each split criterion, could something like this be used instead?

Reply:
If we have a general method for getting the merit for a single distribution of a split criterion, we save verifications of which split criterion was chosen. In my opinion, a better name of the method would be get_merit_single_dist(). What do you think?